plugin/trino-hive/src/main/java/io/trino/plugin/hive/ion/IonPageSourceFactory.java

```diff
@@ -0,0 +1,140 @@
+            ConnectorSession session,
+            Location path,
+            long start,
+            long length,
+            long estimatedFileSize,
+            long lastModifiedTime,
+            Schema schema,
+            List<HiveColumnHandle> columns,
+            TupleDomain<HiveColumnHandle> effectivePredicate,
+            Optional<AcidInfo> acidInfo,
+            OptionalInt bucketNumber,
+            boolean originalFile,
+            AcidTransaction transaction)
+    {
+        if (!ION_SERDE_CLASS.equals(schema.serializationLibraryName())) {
+            return Optional.empty();
+        }
+        checkArgument(acidInfo.isEmpty(), "Acid is not supported for Ion files");
+
+        // Skip empty inputs
+        if (length == 0) {
+            return Optional.of(noProjectionAdaptation(new EmptyPageSource()));
+        }
+
+        if (start != 0) {
+            throw new TrinoException(HIVE_CANNOT_OPEN_SPLIT, "Split start must be 0 for Ion files");
+        }
+
+        List<HiveColumnHandle> projectedReaderColumns = columns;
+        Optional<ReaderColumns> readerProjections = projectBaseColumns(columns);
+
+        if (readerProjections.isPresent()) {
+            projectedReaderColumns = readerProjections.get().get().stream()
+                    .map(HiveColumnHandle.class::cast)
+                    .collect(toImmutableList());
+        }
+
+        TrinoFileSystem trinoFileSystem = trinoFileSystemFactory.create(session);
+        TrinoInputFile inputFile = trinoFileSystem.newInputFile(path, estimatedFileSize);
+
+        // todo: optimization for small files that should just be read into memory
+        try {
+            Optional<Codec> codec = CompressionKind.forFile(inputFile.location().fileName())
+                    .map(CompressionKind::createCodec);
+            CountingInputStream countingInputStream = new CountingInputStream(inputFile.newStream());
+            InputStream inputStream;
+            if (codec.isPresent()) {
+                inputStream = codec.get().createStreamDecompressor(countingInputStream);
+            }
+            else {
+                inputStream = countingInputStream;
+            }
+
+            IonReader ionReader = IonReaderBuilder
+                    .standard()
+                    .build(inputStream);
+            PageBuilder pageBuilder = new PageBuilder(projectedReaderColumns.stream()
+                    .map(HiveColumnHandle::getType)
+                    .toList());
+            List<Column> decoderColumns = projectedReaderColumns.stream()
+                    .map(hc -> new Column(hc.getName(), hc.getType(), hc.getBaseHiveColumnIndex()))
+                    .toList();
+            IonDecoder decoder = IonDecoderFactory.buildDecoder(decoderColumns);
+            IonPageSource pageSource = new IonPageSource(ionReader, countingInputStream::getCount, decoder, pageBuilder);
+
+            return Optional.of(new ReaderPageSource(pageSource, readerProjections));
```

Comment:
Does ReaderPageSource close the IonPageSource?

Reply:
I think ReaderPageSource is confusingly named. IIRC it _has-a_ PageSource it _is-not-a_ PageSource. My understanding is that the calling code directly closes the `IonPageSource` but I will check that.

Reply:
I was correct.